File: scr/P120.R
# Calculate the maximum remainder for a given 'a'
maxRemainder <- function(a) {
  remainders <- numeric(0)  # Vector to store remainders
  max_r <- 0
  n <- 1
  
  while(TRUE) {
    r <- ((a - 1)^n + (a + 1)^n) %% (a^2)
    
    # Check if the remainder is already found
    if (r %in% remainders) {
      break  # Break if the remainder starts repeating
    }
    
    remainders <- c(remainders, r)
    max_r <- max(max_r, r)
    
    n <- n + 2  # Increment by 2 as even 'n' does not produce max remainder
  }
  
  return(max_r)
}

# Main function to calculate the sum of maximum remainders for each 'a' in the range 3 to N
P120 <- function(N) {
  sum(sapply(3:N, maxRemainder))
}

# Measure execution time and call P120
start_time <- Sys.time()
result <- P120(1000)
end_time <- Sys.time()

# Output
cat("Result of P120:", result, "\n")
cat("Execution time:", end_time - start_time, "seconds\n")

File: scr/P120.f
program P120
    implicit none
    integer :: N, result, i
    double precision :: start, finish

    N = 1000
    call cpu_time(start)
    result = 0

    do i = 3, N
        result = result + maxRemainder(i)
    end do

    call cpu_time(finish)
    print *, 'Result of P120: ', result
    print *, 'Execution time: ', finish - start, ' seconds'

contains

    ! Calculate the maximum remainder for a given 'a'
    function maxRemainder(a) result(maxR)
        integer, intent(in) :: a
        integer :: maxR, n, r, remainders(a*a), count, j
        logical :: found

        maxR = 0
        n = 1
        count = 0

        do while (.TRUE.)
            r = mod((a - 1)**n + (a + 1)**n, a**2)

            ! Check if the remainder is already found
            found = .FALSE.
            do j = 1, count
                if (remainders(j) == r) then
                    found = .TRUE.
                    exit
                end if
            end do

            if (found) exit

            count = count + 1
            remainders(count) = r
            maxR = max(maxR, r)

            n = n + 2  ! Increment by 2 as even 'n' does not produce max remainder
        end do
    end function maxRemainder

end program P120
